src/vs/workbench/browser/parts/views/viewsViewlet.ts

```diff
@@ -35,7 +35,21 @@ import { Registry } from 'vs/platform/registry/common/platform';
 
 export abstract class TreeViewsViewletPanel extends ViewletPanel {
 
-	protected tree: WorkbenchTree;
+	private _onDidTabFocus = new Emitter<void>();
+	protected onDidTabFocus = this._onDidTabFocus.event;
+
+	private _tree: WorkbenchTree;
+	protected get tree(): WorkbenchTree { return this._tree; }
+	protected set tree(tree: WorkbenchTree) {
+		this._tree = tree;
+		let tabEntry = true;
+
+		this.disposables.push(this.onDidTabFocus(() => this.focusFirstIfNoFocus()));
+		this.disposables.push(this._tree.onDidClick(() => { tabEntry = false; setTimeout(() => tabEntry = true, 10); }));
+		this.disposables.push(this._tree.onDidFocus(() => { if (tabEntry) { this._onDidTabFocus.fire(); } }));
```

Comment:
I do not think tree consumers should infer tab focus from different events and their order. It would be nice if the tree provides the payload in `onDidFocus` event. Other option would be to focus first visible element on `onDidFocus` if there are no focussed elements. Then we do not need any new events from Tree.

Reply:
> It would be nice if the tree provides the payload in onDidFocus event.

Like in the DOM, if you get a focus event, you don't know where it came from.

> Other option would be to focus first visible element on onDidFocus if there are no focussed elements.

This is what @JacksonKearl did first. I do not think that is correct behavior of a base widget. If you focus the widget, it shouldn't focus any elements be default. This behavior is instance specific.

---

Maybe the real problem here has nothing to do with DOM focus, but with the initial element focus. @JacksonKearl  Shouldn't these instances focus the first element **when** they set the list/tree contents, instead of attempting to react on DOM focus?

Reply:
I think that’s a good idea, I will look into it

Reply:
@sandy081 We can't focus the fist visible element always because there are time when you want to select the whole tree with the mouse, for example alt-clicking in the explorer tree.

Reply:
@joaomoreno 

> I do not think that is correct behavior of a base widget

I agree that the base widget `TreeViewsViewletPanel` should not do it by default. But the base widget can provide this behaviour using a parameter `focusFirstElement` and instances can opt in or not. If most of the views opt in then we can make it default in base view.

I think focussing the first element only when the contents are set will solve the problem completely. It does not work in the following scenario I suppose

- Open file explorer -> This will focus first element
- Open a file -> this will focus & select the file in the explorer
- Open other viewlet, say extensions viewlet
- Close the opened file
- Open file explorer again

Now the focus will be on the tree and not on any element.

@JacksonKearl 

> We can't focus the fist visible element always because there are time when you want to select the whole tree with the mouse, for example alt-clicking in the explorer tree.

Yes agreed and the use case here is to open tree context menu. Then we should also allow focussing the tree through keyboard. I see currently you cannot  focus tree by tab/shift+tab if an element is already focused in the tree.

I do not know what is expected as per accessibility, but my opinion is that it can works as follows while navigating through keyboard

- First element or currently focussed element should be focused when
  - View is focussed by command (programatically). When user presses enter on viewlet icon in activity bar or user executes command to open view. This functionality has to be added in `focus` method.

- If user does tab or shift+tab the sequence should go from tree to first element or currently focussed element or vice versa respectively. Tree should support this.